### ISSUE TITLE
FAILED: libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o 

### ISSUE THREAD

**Issue by michaldaniel** [grindhold/libgtkflow#60]
```
[2/30] Compiling c object 'libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o'
FAILED: libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o 
cc  '-Ilibgtkflow/gtkflow-0.2@sha' '-fdiagnostics-color=always' '-Ilibgtkflow' '-I../libgtkflow' '-Ilibgflow/.' '-I../libgflow/.' '-pipe' '-Wall' '-Winvalid-pch' '-O0' '-g' '-fPIC' '-I/usr/include/glib-2.0' '-I/usr/lib/x86_64-linux-gnu/glib-2.0/include' '-pthread' '-I/usr/include/gtk-3.0' '-I/usr/include/at-spi2-atk/2.0' '-I/usr/include/at-spi-2.0' '-I/usr/include/dbus-1.0' '-I/usr/lib/x86_64-linux-gnu/dbus-1.0/include' '-I/usr/include/gio-unix-2.0/' '-I/usr/include/mirclient' '-I/usr/include/mircommon' '-I/usr/include/mircookie' '-I/usr/include/cairo' '-I/usr/include/pango-1.0' '-I/usr/include/harfbuzz' '-I/usr/include/atk-1.0' '-I/usr/include/pixman-1' '-I/usr/include/freetype2' '-I/usr/include/libpng12' '-I/usr/include/gdk-pixbuf-2.0' '-MMD' '-MQ' 'libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o' '-MF' 'libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o.d' -o 'libgtkflow/gtkflow-0.2@sha/drawinghelper.c.o' -c ../libgtkflow/drawinghelper.c
../libgtkflow/drawinghelper.c: In function ‘append_element’:
../libgtkflow/drawinghelper.c:169:24: error: ‘GTK_STATE_FLAG_DROP_ACTIVE’ undeclared (first use in this function)
     { "drop(active)",  GTK_STATE_FLAG_DROP_ACTIVE }
                        ^
../libgtkflow/drawinghelper.c:169:24: note: each undeclared identifier is reported only once for each function it appears in
../libgtkflow/drawinghelper.c:197:7: warning: implicit declaration of function ‘gtk_widget_path_iter_set_object_name’ [-Wimplicit-function-declaration]
       gtk_widget_path_iter_set_object_name (path, -1, name);
       ^
[7/30] Compiling c object 'libgflow/gflow-0.2@sha/libgflow_gflow-0.2@sha_gflow-simple-source.c.o'
ninja: build stopped: subcommand failed.

```

Build fails with **../libgtkflow/drawinghelper.c:169:24: error: ‘GTK_STATE_FLAG_DROP_ACTIVE’ undeclared (first use in this function)**

Any ida what can be the cause and how to fix it? Any help greatly appreciated.

```
The Meson build system
Version: 0.37.1
Source dir: /home/michal/.local/share/libgtkflow
Build dir: /home/michal/.local/share/libgtkflow/build
Build type: native build
Project name: gtkflow
Native vala compiler: valac (valac 0.34.4)
Native c compiler: cc (gcc 5.4.0)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Native dependency glib-2.0 found: YES 2.48.2
Native dependency gobject-2.0 found: YES 2.48.2
Native dependency gtk+-3.0 found: YES 3.18.9
Library m found: YES
Program g-ir-compiler found: YES (/usr/bin/g-ir-compiler)
Program valadoc found: YES (/usr/bin/valadoc)
Program valadoc found: YES (/usr/bin/valadoc)
```

**Comment by grindhold** [comment]
interesting. maybe porting the drawing methods to Gtk+ >= 3.20 broke compatibility with older Gtk3 versions. Will investigate.

**Comment by grindhold** [comment]
the new implementation of the drawing methods uses GTK_STATE_FLAG_DROP_ACTIVE as the compiler says. this symbol is available only in Gtk3 >= 3.20. I will first try wheter the old GTK_STATE_FLAG_ACTIVE suffices to achieve the same painting behaviour. If not, I will set Gtk+ 3.20 as the minimum dependency version for compiling libgtkflow. Thanks for bringing this to my attention.

TL;DR: should work with a Gtk version > 3.18 .

**Comment by michaldaniel** [comment]
@grindhold It's Ubuntu 16.04 that is still sitting on 3.18 in it's repos. It's huge userbase to give up on when shipping the app. If you can't make it work with 3.18 is there any commit you can point me to that would build with lower GTK version?

**Comment by grindhold** [comment]
try  93f3445 . if that does not work, fall back to b33b896 . those are the two commits that made the biggest changes towards being compatible with the new foreign drawing api of gtk.

**Comment by grindhold** [comment]
@michaeldaniel did it work?

**Comment by grindhold** [comment]
close due to inactivity. if there are any other questions, please reopen.

**Comment by michaldaniel** [comment]
Oh sorry about that. I had to move to another solution and didn't notice the notyfications. I know the issue is closed but I'll update it today with results.

**Comment by schlagenhauf** [comment]
I can confirm that  93f3445 builds on Ubuntu 16.04. Are there news on a fallback solution for GTK <= 3.18? I'd like to use the resizing fix (#62), but i cannot burden my users with compiling GTK from source just to use my application.

**Comment by benwaffle** [comment]
@schlagenhauf 
>but i cannot burden my users with compiling GTK from source just to use my application.

That's where [flatpak](http://flatpak.org) is useful - you can distribute it with the latest version of GTK+, regardless of the version installed on the system

**Comment by grindhold** [comment]
@schlagenhauf if ubuntu 16.04 delivers flatpack, i'd recommend to use it. it would be the most comfortable solution for everyone involved (users, me, you).

If not, there would be the solution to maintain a separate branch for keeping compatibility to Gtk 3.18. But this would not be done by me due to several reasons:
 - I have limited spare time due to a dayjob and a bunch of projects i support in parallel.
 - I myself do not use ubuntu.
 - This library is still < 1.0, so breaking APIs are legal.
However if you want, you can do this. I don't change things that often and the work would mainly consist of cherry-picking commits from master onto the 3.18 branch and see if stuff still works under ubuntu. As you seem to use ubuntu, it would be much easier for you.
If you want to do that i'll happily grant you commit-access.

**Comment by schlagenhauf** [comment]
Thanks @benwaffle , @grindhold , i'll come back to you when i have found a working solution.

**Comment by schlagenhauf** [comment]
Flatpak is currently only available as a PPA for Ubuntu. However it looks like a clean solution, i'll keep it in mind for when i'm reaching a less volatile state in my application.
I'll have a look into maintaining a separate branch and come back to you.

**Comment by schlagenhauf** [comment]
I created a GTK-3.18 compatibility branch: https://github.com/schlagenhauf/libgtkflow/tree/gtk-3.18-compatible

It's not really tested yet, however it compiles and i can use the rescaling fix in my application.

**Comment by grindhold** [comment]
awesome work, @schlagenhauf . i invited you for collaboration and pushed your branch to the repo

**Comment by schlagenhauf** [comment]
Thanks!

**Comment by schlagenhauf** [comment]
Short followup question: What would be the best maintenance policy for the compatibility branch?
* Cherry-picking: Creates artificial copies, cluttered at the front.
* Merging: Requires to separate incompatible parts at ever merge.
* Rebasing: Changes history of a public branch. Looks nice however.

What would you recommend?

**Comment by grindhold** [comment]
in this case i think cherry-picking is the best solution, because you might want to omit certain commits that exclusively work correctly with gtk >= 3.20 and replace them with other ones that work with  gtk <= 3.18. neither rebasing nor merging allow this, as these operations are being executed on whole branches.

i think the first time this will be needed is, when you want to port the "selecting nodes"-feature. you will most likely need to use gtk's old foreign drawing methods [ "foreign-drawing" is gtk-speak for "using graphical elements from the library and draw custom stuff with them without using the widget classes] to get this to work.

for reference, you can look at how drawing nodes is implemented in the 3.18-compatible-branch or you may consult the gtk-3.18-version of the gtk-demo program which comes with libgtk-3-bin. I wish you the best of luck. don't hesitate to ask if you need to know anything.
